### PR TITLE
Update how TM kernels get eigenstrain names for thermal coupling

### PR DESCRIPTION
**Pull request by bwspenc**
ref #14931

### CODE REVIEW

**Comment by bwspenc** [comment]
@dschwen I think this is right now.  This will break BISON, so assuming it passes the tests and you're happy with it, probably best to wait until next week to merge it so I can get that PR put together (which will be easy).

**Comment by bwspenc** [comment]
FYI, @WilkAndy This changes the name of one of the parameters in your actions in porous_flow to be consistent with tensor_mechanics module changes.

**Comment by WilkAndy** [comment]
Thanks @bwspenc - this is one of those things i had forgotten that i ever knew.

**Comment by bwspenc** [comment]
@WilkAndy I'm assuming that the name change won't unduly affect you. At some point, you may want to consider having those Actions in porous_flow use the tensor_mechanics Master action to create the mechanics objects in a way consistent with the other mechanics Actions. We have several other Actions that use it by deriving from it. I don't know if that would make sense for you. If not, you could also have your action create it as a meta-action.

**Comment by WilkAndy** [comment]
> 
> 
> @WilkAndy I'm assuming that the name change won't unduly affect you. At some point, you may want to consider having those Actions in porous_flow use the tensor_mechanics Master action to create the mechanics objects in a way consistent with the other mechanics Actions. We have several other Actions that use it by deriving from it. I don't know if that would make sense for you. If not, you could also have your action create it as a meta-action.

Yea, you're probably right about wanting the PorousFlow Actions to add the TM Master.  I'll put it on my todo list.

a

File: modules/tensor_mechanics/include/kernels/StressDivergenceTensors.h
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/StressDivergenceTensors.h]: in existing parlance this would be `_deigenstrain_dargs`

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]:
```suggestion
    _temp_var(isCoupled("temperature") ? coupled("temperature") : libMesh::invalid_uint),
```

and make `_temp_var` an `unsigned int`

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: I don't get this logic. If `_var.number() == _temp_var` then this shouldn't be off diagonal, or is this special thing with scalar vars?

File: modules/tensor_mechanics/src/kernels/WeakPlaneStress.C
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/WeakPlaneStress.C]: if temperature is not coupled then `jvar` will never be `_temp_var`

**Review by dschwen — APPROVED** [review]
Ok, looks fine.

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C
Comment (bwspenc) [reply modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: Separate instances of this class are set up for the each of the displacements and the temperature, each with a different `variable` parameter. `_var` is that field variable.  `jvar` is the scalar variable. Off-diagonal coupling terms are computed for the interaction between the field var `_var` and scalar var `jvar`. I think this is right.

File: modules/tensor_mechanics/src/kernels/WeakPlaneStress.C
Comment (bwspenc) [reply modules/tensor_mechanics/src/kernels/WeakPlaneStress.C]: So you would just change this to read:
```
if (jvar == _temp_var)
```
correct?  No objections to that, but kind of outside the scope of this PR.

File: modules/tensor_mechanics/include/kernels/StressDivergenceTensors.h
Comment (bwspenc) [reply modules/tensor_mechanics/include/kernels/StressDivergenceTensors.h]: Or when we improve that, maybe `_deigenstrain_dcoupledvars`?

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C
Comment (bwspenc) [reply modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: We need the variable names, though, for the call to `getMaterialPropertyDerivatives()`. How do you get the name from the variable index?  Storing the variable pointer doesn't seem like a problem to me, and it seems consistent with how this is done in other classes that use material property derivatives.

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C
Comment (bwspenc) [reply modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: In looking at this a little more, the only reason we need `temperature` and `displacements` is to figure out what variable `_var` is. We should probably still pass those parameters in to this class, but we could sort all that out in the constructor and store off an enumeration or something that tells is whether we're working with disp_x, disp_y, or temperature. That might make this code easier to understand.

I can take a stab at refactoring this, but I think that should be separate from this PR. I think we need to do more testing on the Jacobians from this to make sure they're right while we're at it.

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C
Comment (dschwen) [reply modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: I've never seen it done that way, but if it works it's fine with me.

**Pull request merged**
